File: PropertyExchange/Presentation/PropertyExchange.Presentation.UI/src/app/shared/models/property.model.ts
import { TenantLeaseDetails, TenantRentPaymentDetailsVM } from "./tenant.model";

export class Property {
    Id: number;
    PropertyID: string;
    Name: string;
    Description: string;
    Location: string;
    PropertyType: string;
    HoldingType:string;
    OverallAreaInSft: number;
    IsActive: boolean;
    IsAvailableForInvesting: boolean;
    InitialTotalValuation: number;
    InitialTokenPrice: number;
    InitialTotalNumberOfTokens: number;
    InitialAvailableNumberOfTokens: number;
    InitialTotalTokensValuation: number;
    CurrentTotalValuation: number;
    CurrentTokenPrice: number;
    CurrentTotalNumberOfTokens: number;
    CurrentAvailableNumberOfTokens: number;
    CurrentTotalTokensValuation: number;
    CreatedDate: string;
    LastUpdatedDate: string;

    AddressDetails: PropertyAddressDetails;
    PurchaseOrSaleExpensesBreakOut: PropertyPurchaseOrSaleExpensesBreakOutVM[];
    Passbook: PropertyPassbook[];
    TradeRecords: PropertyTradeRecords[];
    IncomeBreakOut: PropertyIncomeBreakOut[];
    ExpensesBreakOut: PropertyExpenseBreakOut[];
    ValuationMetrics: PropertyValuationMetrics[];
    ProjectedValuationMetrics: PropertyProjectedValuationMetrics[];
    TenantDetails: TenantDetails[];
}

export class PropertyAddressDetails {
    Id:number;
    PropertyID: string;
    AddressLine1: string;
    AddressLine2: string;
    AddressLine3: string;
    Landmark: string;
    City: string;
    State: string;
    Country: string;
    PINCode: number;
    CreatedDate: string
    LastUpdatedDate: string;
    Property: Property;
}

export class PropertyPurchaseOrSaleExpensesBreakOutVM {
    Id:number;
    PropertyID: string;
    PurchaseOrSale: string;
    ExpenseType: string;
    Description: string;
    ExpenseAmount: number;
    ExpenseIncurredOnDate: string;
    ExpenseMonth: number;
    ExpenseYear: number;
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
}

export class PropertyPassbook
{
    Id:number;
    PropertyID: string;
    TxnNumber : string; 
    TxnType: string;
    Quantity: number;
    Price: number;
    TotalTxnValue:number;
    OrderDateTime: string;
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
}

export class PropertyTradeRecords
{
    Id:number;
    PropertyID: string;
    CurrentTokenPrice: number;
    CurrentTotalNumberOfTokens: number;
    CurrentAvailableNumberOfTokens: number;
    CurrentTotalTokensValuation: string;
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
}

export class PropertyIncomeBreakOut
{
    Id:number;
    PropertyID: string;
    IncomeType: string;
    Description: string;
    IncomeAmount: number;
    IncomeGeneratedOnDate: string;
    IncomeMonth: number;
    IncomeYear: number;
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
}

export class PropertyExpenseBreakOut
{
    Id:number;
    PropertyID: string;
    ExpenseType: string;
    Description: string;
    ExpenseAmount: number;
    ExpenseIncurredOnDate: string;
    ExpenseMonth: number;
    ExpenseYear: number;
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
}

export class PropertyValuationMetrics{
    Id:number;
    PropertyID: string;
    PricePerSft: number;
    TotalSft: number;
    Month: number; 
    Year: number; 
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
}

export class PropertyProjectedValuationMetrics{
    Id:number;
    PropertyID: string;
    PricePerSft: number;
    TotalSft: number;
    Month: number; 
    Year: number; 
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
}

export class TenantDetails{
    Id:number;
    PropertyID: string;
    TenantID: string;
    Name: string;
    Description: string;
    Email: string;
    PhoneNumber: string;
    Country: string;
    IsActive: boolean;
    TenancyType: string;
    CreatedDate: string;
    LastUpdatedDate: string;
    Property: Property;
    LeaseDetails: TenantLeaseDetails[];
    RentPaymentDetails: TenantRentPaymentDetailsVM[];
}

File: PropertyExchange/Core/PropertyExchange.Core.Domain/Models/Property/PropertyModel.cs
using PropertyExchange.Core.Domain.Models.Property;
using PropertyExchange.Core.Domain.Models.User;
using PropertyExchange.Presentation.API.Models.Tenant;

namespace PropertyExchange.Presentation.API.Models.Property
{
    public class PropertyModel
    {
        public int Id { get; set; }
        public required string PropertyID { get; set; }
        public required string Name { get; set; }
        public required string Description{ get; set; }
        public required string Location { get; set; }
        public required string PropertyType { get; set; }
        public required string HoldingType { get; set; }
        public decimal OverallAreaInSft { get; set; }
        public bool IsActive { get; set; }
        //public bool IsAvailableForInvesting { get; set; }
        //public decimal TokenPrice { get; set; }
        //public decimal CurrentTotalPrice { get; set; }
        //public int TotalNumberOfTokens { get; set; }
        //public int AvailableNumberOfTokens { get; set; }
        public bool IsAvailableForInvesting { get; set; }
        public decimal InitialTotalValuation { get; set; } //Will get it only once from PropertyPurchaseOrSaleExpensesBreakout
        public decimal InitialTokenPrice { get; set; } //Will get it only once from PropertyTradeRecords
        public int InitialTotalNumberOfTokens { get; set; } //Will get it only once from PropertyTradeRecords
        public int InitialAvailableNumberOfTokens { get; set; } //Will get it only once from PropertyTradeRecords
        public int InitialTotalTokensValuation { get; set; } //Will get it only once from PropertyTradeRecords
        public decimal CurrentTotalValuation { get; set; } //Will get it from PropertyValuationMetrics
        public decimal CurrentTokenPrice { get; set; } //Will get it from PropertyTradeRecords
        public int CurrentTotalNumberOfTokens { get; set; } //Will get it from PropertyTradeRecords
        public int CurrentAvailableNumberOfTokens { get; set; } //Will get it from PropertyTradeRecords
        public int CurrentTotalTokensValuation { get; set; } //Will get it from PropertyTradeRecords
        public required string CreatedDate { get; set; }
        public string? LastUpdatedDate { get; set; }
        public required PropertyAddressDetailsModel AddressDetails { get; set; }
        public required List<PropertyPurchaseOrSaleExpensesBreakOutModel> PurchaseOrSaleExpensesBreakOut { get; set; }
        public required List<PropertyPassbookModel> Passbook { get; set; }
        public List<PropertyIncomeBreakOutModel>? IncomeBreakOut { get; set; }
        public required List<PropertyExpensesBreakOutModel> ExpensesBreakOut { get; set; }
        public required List<PropertyValuationMetricsModel> ValuationMetrics { get; set; }
        public required List<PropertyProjectedValuationMetricsModel> ProjectedValuationMetrics { get; set; }
        public List<PropertyTradeRecordsModel>? TradeRecords { get; set; }
        public List<TenantModel>? TenantDetails { get; set; }
        public List<UserPassbookModel>? UserPassbook { get; set; }
        public List<UserHoldingModel>? UserHolding { get; set; }

    }

    public enum PropertyType
    {
        Residential = 0, 
        Commercial = 1, 
        Industrial = 2, 
        RawLand = 3,
        SpecialPurpose = 4
    }

    public enum HoldingType
    {
        Owned = 0,
        Leased = 1
    }
}
